File: app/src/main/java/com/divercity/android/features/home/home/adapter/viewholder/QuestionsViewHolder.java
package com.divercity.android.features.home.home.adapter.viewholder;

import android.graphics.drawable.Drawable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.load.DataSource;
import com.bumptech.glide.load.engine.GlideException;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.RequestOptions;
import com.bumptech.glide.request.target.Target;
import com.divercity.android.R;
import com.divercity.android.core.utils.GlideApp;
import com.divercity.android.core.utils.Util;
import com.divercity.android.model.Question;
import com.divercity.android.repository.session.SessionRepository;

import androidx.annotation.Nullable;
import androidx.cardview.widget.CardView;
import androidx.recyclerview.widget.RecyclerView;

public class QuestionsViewHolder extends RecyclerView.ViewHolder {

    private ImageView mImgAuthor;
    private ImageView mImgAnswerAuthor;
    private TextView mTxtGroupName;
    private TextView mTxtQuestion;
    private TextView mTxtAnswer;
    private ImageView mImgGroup;
    private FrameLayout mLayUnseen;
    private TextView mTxtUnseen;
    private TextView mTxtAuthorTime;
    private CardView cardViewImgMainContainer;
    private ImageView imgMainPicture;

    private SessionRepository sessionRepository;
    private Listener listener;

    private QuestionsViewHolder(View itemView,
                                SessionRepository sessionRepository,
                                Listener listener) {
        super(itemView);
        this.sessionRepository = sessionRepository;
        this.listener = listener;
        mTxtGroupName = itemView.findViewById(R.id.item_quest_txt_groupname);
        mImgAnswerAuthor = itemView.findViewById(R.id.item_quest_img_answer);
        mImgAuthor = itemView.findViewById(R.id.item_group_img);
        mTxtQuestion = itemView.findViewById(R.id.item_quest_txt_author_name);
        mTxtAnswer = itemView.findViewById(R.id.item_quest_txt_answer);
        mLayUnseen = itemView.findViewById(R.id.item_quest_lay_unseen);
        mTxtUnseen = itemView.findViewById(R.id.item_quest_txt_unseen);
        mTxtAuthorTime = itemView.findViewById(R.id.item_quest_txt_author_time);
        cardViewImgMainContainer = itemView.findViewById(R.id.item_quest_cardview_pic_main);
        imgMainPicture = itemView.findViewById(R.id.item_quest_img_main);
    }

    public void bindTo(Question data) {
        String urlMain = data.getQuestionPicUrl();
        if (urlMain != null) {
            GlideApp.with(itemView)
                    .load(urlMain)
                    .listener(new RequestListener<Drawable>() {
                        @Override
                        public boolean onLoadFailed(@Nullable GlideException e, Object model, Target<Drawable> target, boolean isFirstResource) {
                            cardViewImgMainContainer.setVisibility(View.GONE);
                            return false;
                        }

                        @Override
                        public boolean onResourceReady(Drawable resource, Object model, Target<Drawable> target, DataSource dataSource, boolean isFirstResource) {
                            cardViewImgMainContainer.setVisibility(View.VISIBLE);
                            return false;
                        }
                    })
                    .into(imgMainPicture);
        } else {
            cardViewImgMainContainer.setVisibility(View.GONE);
        }

        String urlImgAuthor = data.getAuthorProfilePicUrl();
        if (urlImgAuthor != null)
            GlideApp.with(itemView)
                    .load(urlImgAuthor)
                    .apply(new RequestOptions().circleCrop())
                    .into(mImgAuthor);
        try {
            String urlImgAnswerAuthor = data.getLastAnswerAuthorPicture();
            GlideApp.with(itemView)
                    .load(urlImgAnswerAuthor).into(mImgAnswerAuthor);
        } catch (NullPointerException e) {
            GlideApp.with(itemView)
                    .load(sessionRepository.getUserAvatarUrl())
                    .into(mImgAnswerAuthor);
//            mImgAnswerAuthor.setImageResource(R.drawable.tab_profile_inactive);
        }
        mTxtQuestion.setText(data.getQuestion());
        mTxtAuthorTime.setText(" · " + data.getAuthorName() + " · " + Util.getTimeAgoWithStringServerDate(data.getCreatedAt()));
        mTxtGroupName.setText(data.getGroupTitle());
        try {
            mTxtAnswer.setText(data.getLastAnswerAuthorName() + ": " +
                    data.getLastAnswer());
        } catch (NullPointerException e) {
            mTxtAnswer.setText("");
            mTxtAnswer.setHint("Be the first to comment");
        }

        itemView.setOnClickListener(view -> {
            if (listener != null)
                listener.onQuestionClick(data);
        });
    }

    public static QuestionsViewHolder create(ViewGroup parent, SessionRepository sessionRepository, Listener listener) {
        LayoutInflater layoutInflater = LayoutInflater.from(parent.getContext());
        View view = layoutInflater.inflate(R.layout.item_question, parent, false);
        return new QuestionsViewHolder(view, sessionRepository, listener);
    }

    public interface Listener {

        void onQuestionClick(Question question);
    }
}

File: app/src/main/java/com/divercity/android/features/home/home/adapter/viewholder/QuestionsViewHolder.kt
package com.divercity.android.features.home.home.adapter.viewholder

import android.graphics.drawable.Drawable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.load.DataSource
import com.bumptech.glide.load.engine.GlideException
import com.bumptech.glide.request.RequestListener
import com.bumptech.glide.request.RequestOptions
import com.bumptech.glide.request.target.Target
import com.divercity.android.R
import com.divercity.android.core.utils.GlideApp
import com.divercity.android.core.utils.Util
import com.divercity.android.model.Question
import com.divercity.android.repository.session.SessionRepository
import kotlinx.android.synthetic.main.item_question.view.*

class QuestionsViewHolder
private constructor(itemView: View, private val listener: Listener?, val sessionRepository: SessionRepository) :
    RecyclerView.ViewHolder(itemView) {

    fun bindTo(position: Int, data: Question?) {

        data?.let { q ->
            itemView.apply {
                item_quest_txt_groupname.text = q.groupTitle
                item_quest_txt_author_time.text =
                    " · " + data.authorName + " · " +
                            Util.getTimeAgoWithStringServerDate(data.createdAt)
                item_quest_txt_question.text = q.question

                val questionUrlPicture = q.checkedPicture()
                if (questionUrlPicture != null) {
                    GlideApp.with(itemView)
                        .load(questionUrlPicture)
                        .listener(object : RequestListener<Drawable> {
                            override fun onLoadFailed(
                                e: GlideException?,
                                model: Any,
                                target: Target<Drawable>,
                                isFirstResource: Boolean
                            ): Boolean {
                                item_quest_cardview_pic_main.visibility = View.GONE
                                return false
                            }

                            override fun onResourceReady(
                                resource: Drawable,
                                model: Any,
                                target: Target<Drawable>,
                                dataSource: DataSource,
                                isFirstResource: Boolean
                            ): Boolean {
                                item_quest_cardview_pic_main.visibility = View.VISIBLE
                                return false
                            }
                        })
                        .into(item_quest_img_main)
                } else {
                    item_quest_cardview_pic_main.visibility = View.GONE
                }

                GlideApp.with(itemView)
                    .load(q.authorProfilePicUrl)
                    .apply(RequestOptions().circleCrop())
                    .into(item_group_img)

                if (q.lastAnswer != null) {
                    GlideApp.with(itemView)
                        .load(q.lastAnswerAuthorPicture)
                        .into(item_quest_img_answer)

                    item_quest_txt_answer.text =
                        data.lastAnswerAuthorName + ": " + data.lastAnswer

                } else {
                    GlideApp.with(itemView)
                        .load(sessionRepository.getUserAvatarUrl())
                        .into(item_quest_img_answer)

                    item_quest_txt_answer.text = ""
                    item_quest_txt_answer.hint = "Be the first to comment"
                }

                setOnClickListener {
                    listener?.onQuestionClick(position,q)
                }
            }
        }
    }

    interface Listener {

        fun onQuestionClick(position: Int, question: Question)
    }

    companion object {

        fun create(parent: ViewGroup, listener: Listener?, sessionRepository: SessionRepository): QuestionsViewHolder {
            val layoutInflater = LayoutInflater.from(parent.context)
            val view = layoutInflater.inflate(R.layout.item_question, parent, false)
            return QuestionsViewHolder(view, listener, sessionRepository)
        }
    }
}
